test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java

```diff
@@ -125,46 +154,54 @@ public void releasePagesFromHead(int bytesToRelease) {
         return buffers;
     }
 
-    public void incrementIndex(int delta) {
-        int newIndex = delta + internalIndex;
+    public void incrementIndex(long delta) {
+        if (delta < 0) {
+            throw new IllegalArgumentException("Cannot increment an index with a negative delta [" + delta + "]");
+        }
+
+        long newIndex = delta + internalIndex;
         if (newIndex > capacity) {
             throw new IllegalArgumentException("Cannot increment an index [" + internalIndex + "] with a delta [" + delta +
                 "] that will result in a new index [" + newIndex + "] that is greater than the capacity [" + capacity + "].");
         }
         internalIndex = newIndex;
     }
 
-    public int getIndex() {
+    public long getIndex() {
         return internalIndex;
     }
 
-    public int getCapacity() {
+    public long getCapacity() {
         return capacity;
     }
 
-    public int getRemaining() {
+    public long getRemaining() {
         return capacity - internalIndex;
     }
 
-    private int numPages(int capacity) {
-        return (capacity + pageMask) >>> pageShift;
+    private int numPages(long capacity) {
+        final long numPages = (capacity + pageMask) >>> pageShift;
+        if (numPages > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("pageSize=" + (pageMask + 1) + " is too small for such as capacity: " + capacity);
+        }
+        return (int) numPages;
     }
 
-    private int pageIndex(int index) {
-        return index >>> pageShift;
+    private int pageIndex(long index) {
+        return (int) (index >>> pageShift);
     }
 
-    private int indexInPage(int index) {
-        return index & pageMask;
+    private int indexInPage(long index) {
+        return (int) (index & pageMask);
```

Comment:
I usually use `Math.toIntExact()` for this just to have extra protection but I think it's ok here